### PR TITLE
Added a draft for topoloy parsing with netdiff.

### PR DESCRIPTION
**Pull request by gabri94**
What do u think it's just a draft, i have some doubt about the import_module.

Also i had to fetch the node from the links, because i don't have a way to filter the interfaces.

### CODE REVIEW

File: nodeshot/networking/links/models/link.py [hunk 1]
```diff
@@ -33,6 +65,26 @@ class Link(BaseAccessLevel):
...
+                                    related_name='topology that draw this interface', blank=True, null=True,
+                                    help_text=_('mandatory to draw the link dinamically'))
+
+    '''
+    we need a function to export the interfance in netjson format
+    '''
+
+    def to_netjson(self):
+            return {'source': self.interface_a.address.mac,
+                    'target': self.interface_a.address.mac,
```
Comment (PabloCastellano) [inline nodeshot/networking/links/models/link.py]: it should be self.interface_b.address.mac, shouldn't it?

File: nodeshot/core/base/utils.py [hunk 2]
```diff
@@ -112,7 +112,56 @@ def ago(**kwargs):
...
+            f_a = Interface.objects.get(mac=link[0])
+            if_b = Interface.objects.get(mac=link[1])
+            try:
+                l = Link.objects.get(interface_a=if_a, interface_b=if_b)
+                l.delete()
+            except Link.DoesNotExist:
+                pass
+
+
+def exist_node(node, nodes):
```
Comment (PabloCastellano) [inline nodeshot/core/base/utils.py]: It looks very like Javaish to me. :cry: 

I would do instead:

```
def exist_node(id, nodes):
    return any([node for node in nodes if node.get('id') == id])
```

File: nodeshot/core/base/utils.py [hunk 3]
```diff
@@ -112,7 +112,56 @@ def ago(**kwargs):
...
+        classparsr = getattr(module, parser_class_name)
+        njparser = NetJsonParser(to_netjson(topology))
+        parser = classparser(topology.url)
+        graph_diff = diff(njparser, parser)
+        for link in graph_diff['added']:
+            if_a = Interface.objects.get(mac=link[0])
+            if_b = Interface.objects.get(mac=link[1])
+            Link.objects.create(interface_a=if_a, interface_b=if_b, metric_value=link[2]['weight'])
+        for link in graph_diff['removed']:
+            f_a = Interface.objects.get(mac=link[0])
```
Comment (PabloCastellano) [inline nodeshot/core/base/utils.py]: looks like a typo: should it be if_a?

File: nodeshot/core/base/utils.py [hunk 4]
```diff
@@ -112,7 +112,56 @@ def ago(**kwargs):
...
+    return false
+
+
+def to_netjson(topology):
+    nodes = []
+    links = []
+
+    for link in Link.objects.filter(topology=topology):
+        node = {'id': link.inteface_a.mac}
+        if !exist_node(node, nodes):
```
Comment (PabloCastellano) [inline nodeshot/core/base/utils.py]: does it work? should be:
    if not exist_node(...)

**Comment by gabri94** [comment]
corrected it here https://github.com/gabri94/nodeshot/commit/5680d9738d74f8c1566b3edc9c8e1b6c2cea2e50

thanks @PabloCastellano

**Comment by nemesisdesign** [comment]
@gabri94 do `push -f` into your master branch so the corrections will appear here

File: nodeshot/core/base/utils.py (hunk 2 above)
Comment (nemesisdesign) [inline nodeshot/core/base/utils.py]: :+1:

**Comment by nemesisdesign** [comment]
probably netdiff is missing from the requirements file too.

**Comment by nemesisdesign** [comment]
The functions in `nodeshot.core.base.utils` should go into `nodeshot.networking.links.utils` because that's where they belong (they're not used by other modules)

File: nodeshot/networking/links/models/link.py [hunk 5]
```diff
@@ -16,6 +16,38 @@
...
+
+class Topology(BaseDate):
+    name = models.CharField(_('name'), max_length=75, unique=True)
+    slug = models.SlugField(max_length=75, db_index=True, unique=True, blank=True)
+    backend = models.CharField(_('backend'), max_length=128,
+        choices=NETDIFF_PARSERS,
+        help_text=_('select the netdiff's parser to parse the topology from a routing protocol'))
+    url = models.URLField(_('url'))
+    # TODO: find a way to move this in layers
+    if 'nodeshot.core.layers' in settings.INSTALLED_APPS:
```
Comment (nemesisdesign) [inline nodeshot/networking/links/models/link.py]: `nodeshot.networking.links` depends on `nodeshot.core.layers' so the `if` is not needed
reference: https://github.com/ninuxorg/nodeshot/blob/master/nodeshot/networking/links/models/__init__.py

**Pull request merged**